web/src/components/storage/VolumeForm.jsx

```diff
@@ -0,0 +1,478 @@
+      break;
+    case SIZE_METHODS.MANUAL:
+      updatedAttrs = { minSize: size, maxSize: size, fixedSizeLimits: true };
+      break;
+    case SIZE_METHODS.RANGE:
+      updatedAttrs = { minSize, maxSize: formData.maxSize ? maxSize : -1, fixedSizeLimits: true };
+      break;
+  }
+
+  return { ...volume, ...updatedAttrs };
+};
+
+/**
+ * Form-related helper for guessing the size method for given volume
+ *
+ * @param {object} volume - a storage volume object
+ * @return {string} corresponding size method
+ */
+const sizeMethodFor = (volume) => {
+  const { adaptiveSizes, fixedSizeLimits, minSize, maxSize } = volume;
+
+  if (adaptiveSizes && !fixedSizeLimits) {
+    return SIZE_METHODS.AUTO;
+  } else if (minSize !== maxSize) {
+    return SIZE_METHODS.RANGE;
+  } else {
+    return SIZE_METHODS.MANUAL;
+  }
+};
+
+/**
+ * Form-related helper for preparing data based on given volume
+ *
+ * @param {object} volume - a storage volume object
+ * @return {object} an object ready to be used as a "form state"
+ */
+const prepareFormData = (volume) => {
+  const { size: minSize, unit: minSizeUnit } = splitSize(volume.minSize);
+  const { size: maxSize, unit: maxSizeUnit } = splitSize(volume.maxSize);
+
+  return {
+    size: minSize,
+    sizeUnit: minSizeUnit,
+    minSize,
+    minSizeUnit,
+    maxSize,
+    maxSizeUnit,
+    sizeMethod: sizeMethodFor(volume),
+    mountPoint: volume.mountPoint
+  };
+};
+
+/**
+ * Initializer function for the React#useReducer used in the {@link VolumesForm}
+ *
+ * @param {object} volume - a storage volume object
+ * @returns {object} a ready to use initial state
+ */
+const createInitialState = (volume) => {
+  return {
+    volume,
+    formData: prepareFormData(volume),
+    errors: {}
+  };
+};
+
+/**
+ * The VolumeForm reducer
+ */
+const reducer = (state, action) => {
+  const { type, payload } = action;
+
+  switch (type) {
+    case "CHANGE_VOLUME": {
+      return createInitialState(payload.volume);
+    }
+
+    case "UPDATE_DATA": {
+      return {
+        ...state,
+        formData: {
+          ...state.formData,
+          ...payload
+        }
+      };
+    }
+
+    case "SET_ERRORS": {
+      return { ...state, errors: payload };
+    }
+
+    default: {
+      return state;
+    }
+  }
+};
+
+/**
+ * Form used for adding a new file system from a list of templates
+ * @component
+ *
+ * @note VolumeForm does not provide a submit button. It is the consumer's
+ * responsibility to provide both: the button for triggering the submission by
+ * using the form id and the callback function used to perform the submission
+ * once the form has been validated.
+ *
+ * @param {object} props
+ * @param {string} props.id - Form ID
+ * @param {object[]} props.templates - Volume templates
+ * @param {onSubmitFn} props.onSubmit - Function to use for submitting a new volume
+ *
+ * @callback onSubmitFn
+ * @param {object} volume
```

Comment:
np: Although we have defined a [Volume type](https://github.com/openSUSE/agama/blob/58ca603d7e533e46e72d3161c8d722efafca20a9/web/src/client/storage.js#L210), I guess we lost track in some of the previous calls. For someone unfamiliar with the code, it might help to put the proper types. 😅

Reply:
You're completely right. I'll try to fix it.